content/gocourse-2019-05-melbourne.slide

```diff
@@ -813,49 +816,106 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 .caption J. D. Frazer "Illiad", User Friendly - [[https://en.wikipedia.org/wiki/User_Friendly#Crud_Puppy][Crud Puppy]]
 
 
-* Errors (1/3)
+* Errors in Go
 
-"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
-
-The `error` type is a built-in interface:
+- `error` a builtin interface ([[https://golang.org/src/builtin/builtin.go?s=11246:11286#L251][source: error]])
+- Check errors explicitly
+- The [[https://golang.org/pkg/errors/][`errors`]] package has functions for creating new error type
+- Creation of new error type can be done in a few ways:
 
-	type error interface {
-	    Error() string
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
 	}
+- Or use formatted error message using `fmt.Errorf`:
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
+	}
+
+* Errors in Go (continued)
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+- Custom error types can be created. Example from `net` package [[https://golang.org/pkg/net/#AddrError][net/AddrError]]:
+	type AddrError struct {
+		Err  string
+		Addr string
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
+	func (e *AddrError) Error() string {
+		if e == nil {
+			return "<nil>"
+		}
+		s := e.Err
+		if e.Addr != "" {
+			s = "address " + e.Addr + ": " + s
 		}
-		return catStore.cats[id], nil
+		return s
 	}
+- The error values can be logged and printed:
 
+	_, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-* Errors (2/3)
 
-"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
+* Error Handling
 
-Panic should be used when something went unexpectedly wrong.
-For instance, initialise a regular expression with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if the pattern is a string constant.
-`MustCompile` will `panic` at initialisation.
+- Check for the presence of an error (compare with nil)
+- Do not compare error variables ([[https://goplay.space/#mgw9QAmeLUP][samples/errors4.go]])
```

Comment:
I'm not sure if I agree with it. Some packages use a global err variable to signify one specific error which can be compared against, e.g. https://godoc.org/github.com/garyburd/redigo/redis#pkg-variables

seen used in the wild as:
```go
        bytes, err := redis.Bytes(conn.Do("GET", key))
	if err == redis.ErrNil {
		return as.CacheMissErr
	}
	if err != nil {
		return as.ErrorEf(as.CacheErr, err, `Redis GET error %s`, key)
	}
```

Reply:
I'll change it to donot compare two instances of error variables. I hope that clarifies it. Because if you did compare things will break